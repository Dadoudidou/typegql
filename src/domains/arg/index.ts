import { argRegistry } from './registry';
export { compileFieldArgs } from './compiler';
import { ArgOptions, defaultArgOptions } from './options';

export function Arg(options: ArgOptions = {}): ParameterDecorator {
  return (target: Object, fieldName: string, argIndex: number) => {
    const compiledOptions = {
      ...defaultArgOptions,
      ...options,
    }
    argRegistry.set(target.constructor, [fieldName, argIndex], compiledOptions);
  };
}
