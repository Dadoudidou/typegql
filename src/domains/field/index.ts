import { fieldsRegistry, FieldInnerConfig, queryFieldsRegistry } from './registry';

export { FieldInnerConfig, fieldsRegistry, queryFieldsRegistry } from './registry';
export { compileAllFields, compileFieldConfig } from './compiler';
import { compileFieldConfig } from './compiler';
export { isQueryField } from './services';
export { FieldError } from './error';

interface FieldOptions {
  description?: string;
  type?: any;
  name?: string;
}

export function Field(options?: FieldOptions): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    const finalConfig: FieldInnerConfig = {
      property: fieldName,
      name: fieldName,
      ...options,
    };

    fieldsRegistry.set(targetInstance.constructor, fieldName, finalConfig);
  };
}
