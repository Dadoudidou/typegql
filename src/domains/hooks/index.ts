import {
  registerFieldAfterHook,
  registerFieldBeforeHook,
  HookExecutor,
} from './registry';

export {
  fieldAfterHooksRegistry,
  fieldBeforeHooksRegistry,
  HookExecutor,
} from './registry';
export { HookError } from './error';

export function Before(hook: HookExecutor): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    registerFieldBeforeHook(targetInstance.constructor, fieldName, hook);
  };
}

export function After(hook: HookExecutor): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    registerFieldAfterHook(targetInstance.constructor, fieldName, hook);
  };
}

export interface GuardOptions {
  msg: string;
}

export function Guard(
  guardFunction: HookExecutor<boolean>,
  options: GuardOptions,
): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    const hook: HookExecutor = async data => {
      const isAllowed = await guardFunction(data);
      if (isAllowed !== true) {
        throw new Error(options.msg);
      }
    };
    registerFieldBeforeHook(targetInstance.constructor, fieldName, hook);
  };
}

export function createGuard(guardFunction: HookExecutor<boolean>, options: GuardOptions) {
  return (guardOptions: GuardOptions = options): PropertyDecorator => {
    return Guard(guardFunction, guardOptions);
  };
}
