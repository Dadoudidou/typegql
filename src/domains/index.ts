export { Type, compileType } from './type';
export { Field } from './field';
export { Arg } from './arg';
export { Inject, Context, Source, Info } from './inject';
export { registerEnum } from './enum';
