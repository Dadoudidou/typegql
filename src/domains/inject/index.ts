import { InjectorResolver, injectorRegistry } from './registry';
export { injectorRegistry, InjectorsIndex, InjectorResolver } from './registry';

export function Inject(resolver: InjectorResolver): ParameterDecorator {
  return (target: Object, fieldName: string, argIndex: number) => {
    injectorRegistry.set(target.constructor, [fieldName, argIndex], resolver);
  };
}

export const Context = Inject((source, args, context, info) => {
  return context;
});

export const Info = Inject((source, args, context, info) => {
  return info;
});

export const Source = Inject((source, args, context, info) => {
  return source;
});
