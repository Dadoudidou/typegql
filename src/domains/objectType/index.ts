import { compileObjectTypeWithConfig, compileObjectTypeAsInputWithConfig } from './compiler';
import { objectTypeRegistry, inputTypeRegistry } from './registry';

export { compileObjectType } from './compiler';
export { ObjectTypeError } from './error';
export { objectTypeRegistry, inputTypeRegistry } from './registry';

export interface ObjectTypeOptions {
  name?: string;
  description?: string;
}

export function ObjectType(options?: ObjectTypeOptions): ClassDecorator {
  return (target: Function) => {
    const config = { name: target.name, ...options };
    const outputTypeCompiler = () =>
      compileObjectTypeWithConfig(target, config);
    objectTypeRegistry.set(target, outputTypeCompiler);
    
    const inputTypeCompiler = () =>
    compileObjectTypeAsInputWithConfig(target, config);
    inputTypeRegistry.set(target, inputTypeCompiler);
  };
}
