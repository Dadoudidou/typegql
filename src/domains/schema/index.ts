export { schemaRegistry, mutationFieldsRegistry, queryFieldsRegistry } from './registry';
import { schemaRegistry } from './registry';
import { compileSchema } from './compiler';
export { compileSchema } from './compiler';
export { Query, Mutation } from './rootFields';

export function Schema(): ClassDecorator {
  return target => {
    const compiler = () => compileSchema(target);
    schemaRegistry.set(target, compiler);
  };
}
