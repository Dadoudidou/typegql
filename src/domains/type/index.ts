import { registerType } from './registry';

export { objectTypeRegistry, inputTypeRegistry, registerType } from './registry';
export { ObjectTypeError } from './error'

export interface ObjectTypeOptions {
  name?: string;
  description?: string;
}

export function ObjectType(options?: ObjectTypeOptions): ClassDecorator {
  return (target: Function) => {
    const config = { name: target.name, ...options };
    registerType(target, config);
  };
}
