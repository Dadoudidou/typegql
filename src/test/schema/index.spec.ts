import { Query, Schema, compileSchema, ObjectType, Field } from 'domains';
import { graphql, GraphQLObjectType } from 'graphql';

describe('@Schema', () => {
  it('should not allow compiling schema not decorated with @Schema', () => {
    class Foo {}

    expect(() => compileSchema(Foo)).toThrowErrorMatchingSnapshot();
  });

  it('should not allow @Schema without any @Query field', () => {
    @Schema()
    class Foo {}

    expect(() => compileSchema(Foo)).toThrowErrorMatchingSnapshot();
  });

  it('should not allow registering query filed that is not compilable to GraphQLObjectType', () => {
    @Schema()
    class Foo {
      @Query() bar: string;
    }

    expect(() => compileSchema(Foo)).toThrow();
  });

  it('should properly register and resolve schema with query field of type GraphQLObjectType', async () => {
    @ObjectType()
    class PlainType {
      @Field() field: string = 'test';
    }

    @Schema()
    class FooSchema {
      @Query()
      bar(): PlainType {
        return new PlainType();
      }
    }

    const schema = compileSchema(FooSchema);
    const result = await graphql(
      schema,
      `
        {
          bar {
            field
          }
        }
      `,
    );

    expect(result).toEqual({ data: { bar: { field: 'test' } } });
  });
});
