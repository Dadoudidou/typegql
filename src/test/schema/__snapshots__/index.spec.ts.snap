// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Schema should not allow @Schema without any @Query field 1`] = `"@Schema Foo: Schema must have at least one field registered with @Query"`;

exports[`@Schema should not allow compiling schema not decorated with @Schema 1`] = `"@Schema Foo: Schema target must be registered with @Schema"`;
